workflow/controller/operator.go

```diff
@@ -2108,6 +2107,43 @@ func (woc *wfOperationCtx) onNodeComplete(node *wfv1.NodeStatus) {
 	)
 }
 
+// recordNodePhaseChangeEvents creates WorkflowNode Kubernetes events for each node
+// that has changes logged during this execution of the operator loop.
+func (woc *wfOperationCtx) recordNodePhaseChangeEvents(old *wfv1.Nodes, new *wfv1.Nodes) {
+	if !woc.controller.Config.NodeEvents.IsEnabled() {
+		return
+	}
+	oldNodes := *old
+	newNodes := *new
+
+	// Check for newly added nodes; send an event for new nodes
+	for nodeName, newNode := range newNodes {
+		oldNode, exists := oldNodes[nodeName]
+		if exists {
+			if oldNode.Phase == newNode.Phase {
+				continue
+			}
+			if oldNode.Phase == wfv1.NodePending && newNode.Completed() {
+				origPhase := newNode.Phase
```

Comment:
I think you should make a DeepCopy() here, so you do have to modify the node, less risky

Reply:
Reverted to using `DeepCopy` in https://github.com/argoproj/argo-workflows/pull/5531/commits/fd20545f4effc6f1174df55ffb6b62127b0b1730.